morph/cli.py

```diff
@@ -116,6 +131,109 @@ def cmd_count(args):
         print '%d\t%s' % (c, m.show().encode('utf-8'))
 
 
+def cmd_next(args):
+    notes_path = args.notes
+    prio_path = args.prio
+    mizer = MIZERS[args.mizer]
+
+    known = load_db('known')
+
+    candidates = defaultdict(list)
+    with codecs.open(notes_path, 'r', 'utf-8') as f:
+        for line in f.readlines():
+            note = line.strip()
+            text = note.split('\t', 1)[0]
+            unknowns = [m for m in mizer.getMorphemesFromExpr(text)
+                        if m not in known.db]
+            if len(unknowns) == 1:
+                candidates[unknowns[0].show()].append(note)
+
+    with codecs.open(prio_path, 'r', 'utf-8') as f:
+        for line in f.readlines():
+            freq, m = line.strip().split('\t', 1)
+            m_cite = m.split('\t', 1)[0]
+            for cand in candidates[m]:
+                print (u'%s\t%s\t%s' % (freq, m_cite, cand)).encode('utf-8')
+
+
+def cmd_grep(args):
+    pattern_string = args.pattern
+    files = args.files
+    max_count = args.max_count
+    mizer = MIZERS[args.mizer]
+
+    pattern = parse_morpheme(pattern_string.decode('utf-8'))
+
+    count = 0
+    for path in files:
+        with codecs.open(path, 'r', 'utf-8') as f:
+            for line in f:
+                ms = mizer.getMorphemesFromExpr(line)
+                if pattern in ms:
+                    sys.stdout.write(line.encode('utf-8'))
+                    count += 1
+                    if max_count is not None and count >= max_count:
+                        return
+
+
+def clear_locs(db, pred):
+    '''Remove from `db` locations that match `pred`, and morphemes left without locations.'''
+    empty_ms = []
+    for m, locs in db.db.iteritems():
+        locs.difference_update([loc for loc in locs if pred(loc)])
+        if not locs:
+            empty_ms.append(m)
+    for m in empty_ms:
+        del db.db[m]
+
+
+def cmd_sync_known(args):
+    filenames = args.input or ['known.txt']
+    should_merge = args.merge
+
+    external_db = load_db('external', allow_missing=True)
+    if not should_merge:
+        clear_locs(external_db, lambda loc: isinstance(loc, morph.morphemes.Nowhere))
+
+    db_dir = os.path.join(profile_path(), 'dbs')
+    loc = morph.morphemes.Nowhere(maturity=30)  # arbitrary but > maturity threshold
+    for filename in filenames:
+        with codecs.open(os.path.join(db_dir, filename), 'r', 'utf-8') as f:
+            for line in f:
+                line = line.strip()
+                if not line or line.startswith('#'):
```

Comment:
Thanks, all question answered! (The individual commands now went for me from "that certainly sounds useful" -> "it makes sense as a whole"). As long as this is somehow visible for the user there are no problems leaving that in.